src/safeposix/syscalls/fs_calls.rs

```diff
@@ -3,3 +3,322 @@
+            statbuf.st_ino = inodenum;
+
+            //delegate the rest of populating statbuf to the relevant helper
+            match inodeobj {
+                Inode::File(f) => {
+                    Self::_istat_helper(f, statbuf);
+                },
+                Inode::CharDev(f) => {
+                    Self::_istat_helper_chr_file(f, statbuf);
+                },
+                Inode::Dir(f) => {
+                    Self::_istat_helper_dir(f, statbuf);
+                },
+                Inode::Pipe(_) => {
+                    panic!("How did you even manage to refer to a pipe using a path?");
+                },
+                Inode::Socket(_) => {
+                    panic!("How did you even manage to refer to a socket using a path?");
+                },
+            }
+            0
+        } else {
+            -1 //
+        }
+
+    }
+
+    fn _istat_helper(inodeobj: &GenericInode, statbuf: &mut StatData) {
+        statbuf.st_mode = inodeobj.mode;
+        statbuf.st_nlink = inodeobj.linkcount;
+        statbuf.st_uid = inodeobj.uid;
+        statbuf.st_gid = inodeobj.gid;
+        statbuf.st_rdev = 0;
+        statbuf.st_size = inodeobj.size;
+        statbuf.st_blksize = 0;
+        statbuf.st_blocks = 0;
+    }
+
+    fn _istat_helper_dir(inodeobj: &DirectoryInode, statbuf: &mut StatData) {
+        statbuf.st_mode = inodeobj.mode;
+        statbuf.st_nlink = inodeobj.linkcount;
+        statbuf.st_uid = inodeobj.uid;
+        statbuf.st_gid = inodeobj.gid;
+        statbuf.st_rdev = 0;
+        statbuf.st_size = inodeobj.size;
+        statbuf.st_blksize = 0;
+        statbuf.st_blocks = 0;
+    }
+
+    fn _istat_helper_chr_file(inodeobj: &DeviceInode, statbuf: &mut StatData) {
+        statbuf.st_dev = 5;
+        statbuf.st_mode = inodeobj.mode;
+        statbuf.st_nlink = inodeobj.linkcount;
+        statbuf.st_uid = inodeobj.uid;
+        statbuf.st_gid = inodeobj.gid;
+        //compose device number into u64
+        statbuf.st_rdev = makedev(&inodeobj.dev);
+        statbuf.st_size = inodeobj.size;
+    }
+
+    //Streams and pipes don't have associated inodes so we populate them from mostly dummy information
+    fn _stat_alt_helper(&self, statbuf: &mut StatData, inodenum: usize, metadata: &FilesystemMetadata) {
+        statbuf.st_dev = metadata.dev_id;
+        statbuf.st_ino = inodenum;
+        statbuf.st_mode = 49590; //r and w priveliged 
+        statbuf.st_nlink = 1;
+        statbuf.st_uid = DEFAULT_UID;
+        statbuf.st_gid = DEFAULT_GID;
+        statbuf.st_rdev = 0;
+        statbuf.st_size = 0;
+        statbuf.st_blksize = 0;
+        statbuf.st_blocks = 0;
+    }
+
+
+    //------------------FSTAT SYSCALL------------------
+
+    pub fn fstat_syscall(&self, fd: i32, statbuf: &mut StatData) -> i32 {
+        let fdtable = self.filedescriptortable.read().unwrap();
+ 
+        if let Some(wrappedfd) = fdtable.get(&fd) {
+            let fdobj = wrappedfd.read().unwrap();
```

Comment:
"obj" makes it seem like a file object and not a descriptor, lets have it be like "filedesc"